lightning/src/events/bump_transaction.rs

```diff
@@ -231,3 +267,486 @@ pub enum BumpTransactionEvent {
+/// [`CoinSelection`] attempt.
+#[derive(Clone, Debug)]
+pub struct Utxo {
+	/// The unique identifier of the output.
+	pub outpoint: OutPoint,
+	/// The output to spend.
+	pub output: TxOut,
+	/// The upper-bound weight consumed by the corresponding input's full witness required to
+	/// satisfy the output's script.
+	pub witness_weight: u64,
+}
+
+/// The result of a successful coin selection attempt for a transaction requiring additional UTXOs
+/// to cover its fees.
+pub struct CoinSelection {
+	/// The set of UTXOs (with at least 1 confirmation) to spend and use within a transaction
+	/// requiring additional fees.
+	confirmed_utxos: Vec<Utxo>,
+	/// An additional output tracking whether any change remained after coin selection. This output
+	/// should always have a value above dust for its given `script_pubkey`.
+	change_output: Option<TxOut>,
+}
+
+/// An abstraction over a bitcoin wallet that can perform coin selection over a set of UTXOs and can
+/// sign for them. The coin selection method aims to mimic Bitcoin Core's `fundrawtransaction` RPC,
+/// which most wallets should be able to satisfy. Otherwise, consider implementing [`WalletSource`],
+/// which can provide a default implementation of this trait when used with [`Wallet`].
+pub trait CoinSelectionSource {
+	/// Performs coin selection of a set of UTXOs, with at least 1 confirmation each, that are
+	/// available to spend. Implementations are free to pick their coin selection algorithm of
+	/// choice, as long as the following requirements are met:
+	///
+	/// 1. `must_spend` contains a set of [`Input`]s that must be included in the transaction
+	///    throughout coin selection.
+	/// 2. `must_pay_to` contains a set of [`TxOut`]s that must be included in the transaction
+	///    throughout coin selection.
+	/// 3. Enough inputs must be selected/contributed for the resulting transaction (including the
+	///    inputs and outputs noted above) to meet `target_feerate_sat_per_1000_weight`.
+	///
+	/// Implementations must take note that [`Input::witness_weight`] only tracks the weight of the
+	/// input's witness. Some wallets, like Bitcoin Core's, may require providing the full input
+	/// weight. Failing to do so may lead to underestimating fee bumps and delaying block inclusion.
+	///
+	/// The `claim_id` uniquely identifies a claim, and should be used to assign a set of UTXOs to
+	/// the claim, such that they can be re-used within new fee-bumped iterations of the original
+	/// claiming transaction, ensuring that claims don't double spend each other. If a specific
+	/// `claim_id` has never had a transaction associated with it, and all of the available UTXOs
+	/// have already been assigned to other claims, implementations must be willing to double spend
+	/// their UTXOs. The choice of which UTXOs to double spend is left to the implementor, e.g., an
+	/// implementation could choose to re-assign the UTXOs assigned to the lowest value claim.
```

Comment:
I think lowest value claim after deduction of the absolute fee, as all claims are not equal in weight? Can we document more heuristics for UTXOs re-assignment, like doing arbitrage based on near-timelock expiration ?